### ISSUE TITLE
Bug in ID mapping: Primary Uniprot IDs are never added during Secondary ID mapping

### ISSUE THREAD

**Issue by CPerscheid** [saezlab/pypath#151]
**Describe the bug**
Primary uniprot IDs are removed during secondary ID mapping

The method util.mappings.Mapper.primary_uniprot() attempts to map secondary Uniprot IDs to primary ones. If it already is a primary one, it should remain in the set of primaries.

However, the if-statement is false: it assumes that self.map_name() returns None if no mapping was found. Instead, this method returns set(). By this, the primary id is never added in the else-branch.

So, instead of 
```
if primary:
     primaries.update(primary)
else:
     # most probably this UniProt is already primary
     primaries.add(uniprot)
```
it should be 
```
if len(primary) > 0:
     primaries.update(primary)
else:
     # most probably this UniProt is already primary
     primaries.add(uniprot)
```

**Comment by deeenes** [comment]
Hi,

The two pieces of code above are equivalent as an empty set evaluates to `False`.
Using the method `primary_uniprot` gives the expected result:

- With a primary ID:
```
from pypath.utils import mapping
mapping.init()
mapping.mapper.primary_uniprot({'P00533'})
# {'P00533'}
```

- With a secondary ID:
```
from pypath.utils import mapping
mapping.init()
mapping.mapper.primary_uniprot({'R9QAJ4'})
# {'Q96J86'}
```

Note: you need to know the taxon of the secondary UniProt ID, if you pass an UniProt from some other taxon to this method it will just return it unchanged.

I hope this helps.

Best wishes,

Denes

**Comment by CPerscheid** [comment]
Hi Denes,

thanks for the fast response. I tried the original code again and also a minimum working example, and it seems you are right (this is one of the moments when you think you found the Error after debugging in the sources the whole afternoon :D ).

In retrospective, I suppose my error occurred because PyPath had not completely initialized the Uniprot database for mapping (because I interrupted the run prior) but the cache said it was there. Deleting the cache was probably the actual solution here.

Well, at least I now have a quite good understanding what happens at PyPath under the hood *haha*
